OADP backup-restore single VM (#86693)
openshift/openshift-docs: modules/oadp-restore-single-vm-from-multiple-vm-backup.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-kubevirt.adoc

:_mod-docs-content-type: PROCEDURE
[id="oadp-restore-single-vm-from-multiple-vm-backup_{context}"]

> = Restoring a single VM from a backup of many VMs

Reviewer: ```suggestion
= Restoring a single VM from a backup of multiple VMs
```

> If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored and the restored VM is not stuck in a `Provisioning` status, use both, the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, you can use the UID of the VM.

Reviewer 2: ```suggestion
If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both, the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, you can use the UID of the VM.
```

Pedantic - they are two independent clauses

> If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, you can use the UID of the VM.

Reviewer 3: I think if you backup with  matchLabels , you do not need the created by
but would like to verify

> If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, you can use the UID of the VM.

Reviewer 3: ```suggestion
If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, you can use the UID of the VM's DataVolume .
```

> If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the Persistent Volume Claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, use the UID of the `DataVolume` of the VM.

Reviewer: ```suggestion
If you have a backup containing many virtual machines (VMs), and you want to restore only one VM, you can use the `LabelSelectors` section in the `Restore` CR to select the VM to restore. To ensure that the persistent volume claim (PVC) attached to the VM is correctly restored, and the restored VM is not stuck in a `Provisioning` status, use both the `app: <vm_name>` and the `kubevirt.io/created-by` labels. To match the `kubevirt.io/created-by` label, use the UID of `DataVolume` of the VM.
```

- Remove "the" or mention the type of `DataVolume`, for example "the `DataVolume` CR" or "the `DataVolume` volume"

.Prerequisites

* You have installed the {oadp-short} Operator.
* You have labeled the VMs that need to be backed up.

> * You have a back up of many VMs.

Reviewer 2: ```suggestion
* You have a backup of many VMs.
```

> * You have a backup of many VMs.

Reviewer: ```suggestion
* You have a backup of multiple VMs.
```

- IMO, "multiple" would be more suitable here, up to you though!

.Procedure

> . Before you take a back up of many VMs, ensure that the VMs are labeled by running the following command:

Reviewer 2: ```suggestion
. Before you take a backup of many VMs, ensure that the VMs are labeled by running the following command:
```

+
[source, terminal]
----

> $ oc label pvc <pvc_name> vm-name=<vm_name> -n openshift-adp 

Reviewer 3: ```suggestion
$ oc label vm  <vm_name> vm-name=<vm_name> -n openshift-adp 
```

> $ oc label vm <vm_name> vm-name=<vm_name> -n openshift-adp 

Reviewer 4: Is this step necessary? I don't see this label being used in the restore.

Reviewer 3: suppse to be app=<vm_name> actually, my mistake

Reviewer 4: Wouldn't this also be needed for the single VM backup doc?

Reviewer 3: yes, your right!

----

. Configure the label selectors in the `Restore` CR as shown in the following example:
+
.Example `Restore` CR
[source,yaml]
----
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: singlevmrestore
  namespace: openshift-adp
spec:
  backupName: multiplevmbackup
  restorePVs: true
  LabelSelectors:
    - matchLabels:

>         kubevirt.io/created-by: <datavolume_uid> # <1>

Reviewer 4: Is this created-by label also needed for the single-VM backup above?

Reviewer 3: yes, so it restores the PVCs
and this is the reason i think RIA should handle it (it currently only restores the datavolume associate with the vm and not the pvc

Reviewer 4: Yeah, I don't think the RIA functionality will override the include/exclude filtering for the restore.

    - matchLabels:
        app: <vm_name> # <2>
----

> <1> Specify the UID of the `DataVolume` of the VM that is to be restored. For example, `b6...53a-ddd7-4d9d-9407-a0c...e5`.

Reviewer: ```suggestion
<1> Specify the UID of `DataVolume` of the VM that you want to restore. For example, `b6...53a-ddd7-4d9d-9407-a0c...e5`.
```
- Suggested a change to make it an active voice.

> <2> Specify the name of the VM to be restored. For example, `test-vm`.

Reviewer: ```suggestion
<2> Specify the name of the VM that you want to restore. For example, `test-vm`.
```

> . To restore, run the following command:

Reviewer: ```suggestion
. To restore a VM, run the following command:
```

Does this instruction mean to restore the VM?

+
[source, terminal]
----
$ oc apply -f <restore_cr_file_name> # <1>
----

> <1> Specify the name of the restore CR file.

Reviewer: ```suggestion
<1> Specify the name of the `Restore` CR file.
```

+
[source, terminal]
----

> $ oc label vm <vm_name> vm-name=<vm_name> -n openshift-adp 

Reviewer 3: ```suggestion
$ oc label vm <vm_name> app=<vm_name> -n openshift-adp 
```
